### DISCUSSION TITLE
Structure snippets in multiple files

### DISCUSSION THREAD

**Discussion by Felipe-9** [L3MON4D3/LuaSnip#1236] [Q&A] [answered]
Previously i had a very (3000 lines long) snippets file for `tex` files and i wanted to organize so that it would be easier to manage and expand.

### here is an example of a tree:
```
/Users/felipepinto/.config/nvim/lua
├── plugins
├── profiles
└── snippets
    ├── init.lua
    └── latex
        ├── G
        │   ├── G1.lua
        │   ├── G2.lua
        │   ├── G3.lua
        └── init.lua
```

### every subfile would return a luatable of snippets, eg:
```lua
local ls = require("luasnip")
local s = ls.snippet     -- build snippets
local fmt = require("luasnip.extras.fmt").fmta -- formatting with [[]] and delimiters=<>
return {
  s({desc="desc",trig="trig"},fmt([[...]],{...})),
  ...
}
```

how to group everything and load without overwritting existing snippets?
if possible it would be ideal for `snippets/init.lua` to load `snippets/latex/init.lua` from which it load all the latex snippets.

im not sure if im doing it wright so far...

### `luasnip.lua` has:
```lua
...
require("snippets.init")
...
```

### `snippets/init.lua` is simply:
```lua
require('snippets.latex.init')
```

### and `snippets/latex/init.lua` is:
```lua
local ls = require("luasnip")
local G = {
  require("snippets.latex.G.G1"),
  require("snippets.latex.G.G2"),
  require("snippets.latex.G.G3"),
}
ls.add_snippets("tex",G)
```

i dont get errors but my snippets also dont load

**Comment by iMagUdspEllr** [comment]
I'm having the same issue. I also can't figure out how to split snippets into multiple files. I was able to write functions that return snippets in separate files and then import them in other files. Unfortunately, the add_snippets function call overwrites other add_snippets function calls. So, only one instance of add_snippets seems to be kept for each file type. But, I think I either don't fully understand LuaSnip or Lua to know how to proceed.

**Reply by Felipe-9** [reply]
At some point i tried the exact same thing, each file would call `ls.add_snippets("tex",...)`, no errors would appear, except, my snippets would be missing (the overriding thing appears). That is actually the reason as why there is a single `snippets/latex/init.lua` file, its the only one to call `ls.add_snippets("tex",...)` for `tex` files.

I still couldnt figure out how to add the snippets imported from the other files...i think it might be interpretting it wrong cause the snippets are inside their own object `{...}` and maybe we need to `flatten` it, but i dont know how to do this

**Reply by iMagUdspEllr** [reply]
I think what you're doing could be almost there. But, your code (in my eyes) means that you're defining a table G that is importing tables. The add_snippets function is meant to accept a table of snippets, not a table of tables of snippets. But, I tried to export only a single snippet and access just the snippet and that didn't work either.

I'm not sure why an imported snippet doesn't work.

**Comment by L3MON4D3** [comment]
Have you tried using the lua-loade? Check the doc, the files can be structure almost exactly as you desire and it hast some nice propertiesike hot reload and jumping to snippet-definition

**Reply by iMagUdspEllr** [reply]
Thank you. I will learn this and report back.

**Reply by Felipe-9** [reply]
i have passed trough the `lua-loaders` from the docs, but im still quite confused how to implement it....
- should we used the `ls.setup` describing `snip_env` on every file that we want to describe snippets?
- Every snippet should be on their own table? in the example `~/snippets/c.lua` like this (for `tex`)?
```lua
return {
      s( -- parabolas
        {
          desc = "parametrizando parabolas",
          trig = "\\formula-parabolas",
        },
        fmt([[
      y=\alpha\,x^2+\beta\,x+\gamma
      \implies &\\&
      \implies
      \begin{cases}
        a = (4\,\alpha)^{-1}
        \\
        x_0
        = -\beta\,2\,a
        = -\beta/2\,\alpha
        \\
        y_0
        = \gamma-x_0^2/4\,a
        = \gamma-\beta/4\,\alpha
      \end{cases}
    ]], {})
      )
    },
    {
      s( -- elipses
        {
          desc = "caracterizando elipses",
          trig = "\\formua-elipses",
        },
        fmt([[
      \frac{(x-x_0)^2}{r_1^2}
      + \frac{(y-y_0)^2}{r_2^2}
      = 1
      &\\&
      \begin{cases}
        c = \myvert{\vec{p_0\,f}}
        = \sqrt{r_1^2+r_2^2}
        \\
        f_1\in\mathbb{r}^2:f_1=(x_0-c,y_0)
        \\
        f_1\in\mathbb{r}^2:f_1=(x_0-c,y_0)
      \end{cases}
    ]], {})
      ),
    }
```
- i dont see any language specification on the files is that obligatory? we cant have many files describing snippets to a single language?
- i see the command `require("luasnip.loaders.from_lua").load({paths = "~/snippets"})` specifying a `path` to which we pass a directory, what if that path has sub dirs, will they be read? can we specify a file instead of a directory? should it be called multiple times for each file we are describing? or should we call it only a single time?

`lua-loaders` might be a solution but im having some trouble understanding how everything will work together or how to implement it

**Reply by L3MON4D3** [reply]
- `snip_env` is defined once and its keys become global variables in the loaded lua-files
- the file should return two tables, the first containing all snippets, the second all autosnippets (essentially, these are passend as-is to add_snippets)
- you can have mutliple files via `snippets/latex/1.lua`, `snippets/latex/2.lua`, etc.
- you should only pass the root-path, the loader descends into subdirectories. So for the latex-files above, pass the path to `snippets/`

Hope that helped :)

**Reply by Felipe-9** [reply]
Alright thankyou very much so if i understand right, this is how we should implement it...

this configuration should work right? the problem is that it isnt working... what is wrong? i dont see any errors, only that my snippets are missing

other than begin in inside a dir named `latex` i dont see anyway the loader knows which language to attach this snippets... how the loader know that `tex` files should have all the snippets in this dir?

<details>
<summary>

### snippets tree

</summary>

this is how i should organize my snippets files for latex...
```
lua/snippets
└── latex
    ├── FCT
    │   ├── 01G.lua
    │   ├── 02G.lua
    │   ├── 03G.lua
    │   ├── 04G.lua
    │   ├── 05G.lua
    │   ├── 06G.lua
    │   ├── 07G.lua
    │   ├── 08G.lua
    │   ├── 09G.lua
    │   └── 10G.lua
    ├── constants.lua
    ├── math.lua
    ├── mysnippets.lua
    ├── templates
    │   ├── mainfile.lua
    │   └── subfile.lua
    └── tikz-pfg.lua

4 directories, 16 files
```
</details>
<details>

<summary>

### `snippets/latex/mysnippets.lua`
</summary>

Showing this as example, but each `file.lua` inside latex follows the same template...
```lua
local ls = require("luasnip")

local s = ls.snippet     -- build snippets
-- local t = ls.text_node -- insert text
local i = ls.insert_node -- user input
local c = ls.choice_node -- multiple options

local extras = require("luasnip.extras")
-- local rep = extras.rep -- repeat for multiple cursors
local ne = extras.nonempty
local fmt = require("luasnip.extras.fmt").fmta -- formatting with [[]] and delimiters=<>

return {
  s(
    {
      desc = "desc",
      trig = "\\trig"
    },
    fmt([[
      ...
      \my_cute_snippet
      <var>
      ...
    ]],{ var = i(1,"my variable") })
  ),
  s(...),
  s(...),
  s(...),
  s(...),
}
```

</details>

<details>
<summary>

### `plugins/luasnip.lua`

</summary>

```lua
return {
  "L3MON4D3/LuaSnip",
  -- follow latest release.
  -- version = "v2.*", -- Replace <CurrentMajor> by the latest released major (first number of latest release)
  -- install jsregexp (optional!).
  build = "make install_jsregexp",
  event = "InsertEnter",
  dependencies = {
    "saadparwaiz1/cmp_luasnip", -- for autocompletion
    "rafamadriz/friendly-snippets", -- useful snippets
  },
  config = function()
    local ls = require("luasnip")

    ls.setup({
      snip_env = {
        s = function (...)
          local snip = ls.s(...)
          table.insert(getfenv(2).ls_file_snippets, snip)
        end,
        parse = function (...)
          local snip = ls.parser.parse_snippet(...)
          table.insert(getfenv(2).ls_file_snippets, snip)
        end
      }
    })

    -- loads vscode style snippets from installed plugins (e.g. friendly-snippets)
    require("luasnip.loaders.from_vscode").lazy_load()
    require("luasnip.loaders.from_lua").load({paths = "snippets"})
    -- require("snippets.init")

    vim.keymap.set({ "i" }, "<C-K>", function()
      ls.expand()
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-k>", function()
      if ls.expand_or_jumpable() then
        ls.expand_or_jump()
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-j>", function()
      if ls.jumpable(-1) then
        ls.jump(-1)
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-E>", function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-n>", function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end)
  end,
}
```

</details>

**Reply by L3MON4D3** [reply]
Oh, afaik latex-files have the tex-filetype, so should `latex/` maybe be `tex/`? Also, the lua-loader only checks the filetype-dir for lua-files, so `latex/a/b.lua` would not be loaded, but `latex/a.lua` should be.
Apart from that, enable detailed log via `ls.log.set_loglevel("info")` (before the call to `load`) and check it after loading via `ls.log.open()`

**Reply by Felipe-9** [reply]
Ok ok, i think i got it this time, here another proposition

- `luasnip.ls` remains as last shown
- since `snippets/lua/data/latex/*` is not written there wont be conflicts, and all latex files will be passed trough `snippets/tex.lua`

this still doesnt show any snippets nor errors, what am i doing wrong?

`ls.log.set_loglevel("info")` and `ls.log.open()` only showed `%a + "filename.tex" line 17`

<details>
<summary>

## tree
</summary>

```
lua/snippets
├── data
│   └── latex
│       ├── G
│       │   ├── G1.lua
│       │   ├── G2.lua
│       │   ├── G3.lua
│       │   ├── G4.lua
│       │   ├── G5.lua
│       │   ├── G6.lua
│       │   ├── G7.lua
│       │   ├── G8.lua
│       │   └── G9.lua
│       ├── constants.lua
│       ├── latex.jsonc
│       ├── math.lua
│       ├── mysnippets.lua
│       ├── templates
│       │   ├── mainfile.lua
│       │   └── subfile.lua
│       └── tikz-pfg.lua
└── tex.lua

5 directories, 18 files
```
</details>
<details>
<summary>

## `snippets/data/latex/example.lua`
</summary>

i figured out that we could `return` the snippets without putting them in an object (without using `{}`)
```lua
local ls = require("luasnip")

local s = ls.snippet -- build snippets
local fmt = require("luasnip.extras.fmt").fmta -- formatting with [[]] and delimiters=<>

return
--
    s(...),
    s(...),
    ...
```
</details>
<details>
<summary>

## `snippets/latex.lua`
</summary>
just groups every latex file and returns

```lua
return {
  require("snippets.data.latex.constants"),
  require("snippets.data.latex.math"),
  require("snippets.data.latex.mysnippets"),
  require("snippets.data.latex.tikz-pfg"),
  -- G
  require("snippets.data.latex.G.G1"),
  require("snippets.data.latex.G.G2"),
  require("snippets.data.latex.G.G3"),
  require("snippets.data.latex.G.G4"),
  require("snippets.data.latex.G.G5"),
  require("snippets.data.latex.G.G6"),
  require("snippets.data.latex.G.G7"),
  require("snippets.data.latex.G.G8"),
  require("snippets.data.latex.G.G9"),
  -- Templates
  require("snippets.data.latex.templates.mainfile"),
  require("snippets.data.latex.templates.subfile"),
}
```
</details>

**Reply by L3MON4D3** [reply]
There should be more in the log, are you sure `set_loglevel` is called before the call to `load`?
I also suspect the path passed to load may not be correct, could you try passing the full path to the `snippets`-directory?

I think making it work using the framework given by the loader is a good idea, if you don't like that you can't nest directories deeper, I'm open to making the maximum depth configurable :)

**Reply by Felipe-9** [reply]
## path on `luasnip.lua`
```lua
...
require("luasnip.loaders.from_lua").load({paths="~/.config/nvim/lua/snippets"})
...
```

## `set_loglevel` 
its a boomer but thats all i see really
<img width="1000" alt="Screenshot 2024-10-01 at 19 36 41" src="https://github.com/user-attachments/assets/99ef3897-6526-4acd-a14a-930cad40258b">

## on nesting directories
i was doing that just so that i would have control (`tex.lua` file to declare which files to include) i tough it would be the same as having the loader decide on the files...

if you are willing to make that configurable, i guess it could be a solution, but im not fully sure if its the problem, you see, i tried to compile all my files into a single (4000 lines) file (`tex.lua`) and imported my snippets both using `ls.add_snippets("tex",...)` and the loader, and at the start i saw some errors (on files like `g10.lua`) from snippets i made mistakes that after fixing my latex files would open without errors, but my snippets would still be missing... at this point i think there might be another problem that isnt communicating right between luasnip and completions or something of the sort

having errors on files tells me that lua compiled the snippet files, but still havent been able to add to snippets

**Reply by L3MON4D3** [reply]
Oh no, I think you're misunderstanding: `ls.log.set_loglevel("info")` has to be called in your luasnip-config, before the call to `load` is made (so that luasnip logs all the info-messages from the lua-loader).
You can then, after loading, in the interactive session, call `ls.log.open()` to view the log :)

**Reply by Felipe-9** [reply]
oh i see, in the config should that line be placed in a specific place? like at the very start or after calling `ls.setup({...})`? :)

<details>
<summary> like this? </summary>

```lua
return {
  "L3MON4D3/LuaSnip",
  -- follow latest release.
  -- version = "v2.*", -- Replace <CurrentMajor> by the latest released major (first number of latest release)
  -- install jsregexp (optional!).
  build = "make install_jsregexp",
  event = "InsertEnter",
  dependencies = {
    "saadparwaiz1/cmp_luasnip", -- for autocompletion
    "rafamadriz/friendly-snippets", -- useful snippets
  },
  config = function()
    local ls = require("luasnip")

    ls.log.set_loglevel('info')

    ls.setup({
      snip_env = {
        s = function (...)
          local snip = ls.s(...)
          table.insert(getfenv(2).ls_file_snippets, snip)
        end,
        parse = function (...)
          local snip = ls.parser.parse_snippet(...)
          table.insert(getfenv(2).ls_file_snippets, snip)
        end
      }
    })


    -- loads vscode style snippets from installed plugins (e.g. friendly-snippets)
    require("luasnip.loaders.from_vscode").lazy_load()
    require("luasnip.loaders.from_lua").load({path="~/.config/nvim/lua/snippets"})
    -- require("snippets.tex")

    vim.keymap.set({ "i" }, "<C-K>", function()
      ls.expand()
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-k>", function()
      if ls.expand_or_jumpable() then
        ls.expand_or_jump()
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-j>", function()
      if ls.jumpable(-1) then
        ls.jump(-1)
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-E>", function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-n>", function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end)
  end,
}
```
</details>

**Reply by Felipe-9** [reply]
### update
i tried putting it everywhere, dont see any difference when calling `ls.log.open()`

Also i noticed how my snippets were giving an error when they had multiple triggers, i noticed the issue #1061 and is `ms` the only easy way to add multiple triggers to snippets? is there any issue with efficiency if i refactor all my snippets to `ms`?

**Reply by L3MON4D3** [reply]
That sounds correct.. very weird, could you manually check whether the file at `vim.fn.stdpath("log")`/luasnip.log contains anything?

`ms` is pretty cool, prefer it if you want to give one snippet multiple triggers :) (it does not duplicate as much data)

**Comment by Felipe-9** [comment, marked as answer]
Hello! its been some time, i finnally found a solution for this.

the solutions i found whent on the path of adding sinppets using `ls.add_snippets('tex',{...})` method.

The reason it wasnt working before is because we were inserting a table of tables instead of having a single table of snippets which was expected by the function.

so with the following loop we can join together tables and insert then into luasnip:
```lua
local snippet_tables = {
  require('snippets.latex.mysnipfile'),
  ...
}

local snippets = {}
for _,sniptable in ipairs(snippet_tables) do
  for _,snip in ipairs(sniptable) do
    table.insert(snippets,snip)
  end
end

ls.add_snippets('tex',snippets)
```

<details><summary>

## luasnip.lua
</summary>

```lua
return {
  "L3MON4D3/LuaSnip",
  build = "make install_jsregexp",
  event = "InsertEnter",
  dependencies = {
    "saadparwaiz1/cmp_luasnip", -- for autocompletion
    "rafamadriz/friendly-snippets", -- useful snippets
  },
  config = function()
    local ls = require("luasnip")

    ls.log.set_loglevel("info")

    -- loads vscode style snippets from installed plugins (e.g. friendly-snippets)
    require("luasnip.loaders.from_vscode").lazy_load()
    require("snippets.init")

    vim.keymap.set({ "i" }, "<C-K>", function()
      ls.expand()
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-k>", function()
      if ls.expand_or_jumpable() then
        ls.expand_or_jump()
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-j>", function()
      if ls.jumpable(-1) then
        ls.jump(-1)
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-E>", function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end, { silent = true })
    vim.keymap.set({ "i", "s" }, "<C-n>", function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end)
  end,
}
```
</details>

## snippets.init
```lua
require("snippets.latex.init")
```

## snippets.latex.init
```lua
local ls = require'luasnip'

local snippet_tables = {
  require("snippets.latex.constants"),
  require("snippets.latex.math"),
  require("snippets.latex.mysnippets"),
  require("snippets.latex.tikz-pfg"),
  -- G
  require("snippets.latex.G.1"),
  require("snippets.latex.G.2"),
  require("snippets.latex.G.3"),
  require("snippets.latex.G.4"),
  require("snippets.latex.G.5"),
  require("snippets.latex.G.6"),
  require("snippets.latex.G.7"),
  require("snippets.latex.G.8"),
  require("snippets.latex.G.9"),
  -- Templates
  require("snippets.latex.templates.mainfile"),
  require("snippets.latex.templates.subfile"),
}

local snippets = {}
for _,sniptable in ipairs(snippet_tables) do
  for _,snip in ipairs(sniptable) do
    table.insert(snippets,snip)
  end
end

ls.add_snippets('tex',snippets)
```

**Reply by L3MON4D3** [reply]
Ah nice, I'm happy you found a solution that works for you :)
